# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/serviceworkers/ServiceWorkerLifetimeExtension.cpp
# Commit: fb720435215f
# Full Hash: fb720435215fa5be3b2bd2101f3a54382954b70c
# Author: Andrew Sutherland <asutherland@asutherland.org>
# Date: 2024-10-24 09:44:34
# Regressor Bug: 1544232
# File Overlap Count: 4
# Description:
#   Bug 1544232 - Limit lifetime extension of SWs by SWs to the sender's lifetime. r=edenchuang
#   
#   This patch introduces an explicit concept of lifetimes with mechanisms
#   in place so that actions taken by Clients (windows or non-ServiceWorker
#   orkers) will extend the lifetime of a ServiceWorker, but a ServiceWorker
# ==============================================================================

diff -r 25452f6623a0 -r fb720435215f dom/serviceworkers/ServiceWorkerLifetimeExtension.cpp
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/dom/serviceworkers/ServiceWorkerLifetimeExtension.cpp	Thu Oct 24 03:02:42 2024 +0000
@@ -0,0 +1,27 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=8 sts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "mozilla/dom/ServiceWorkerLifetimeExtension.h"
+
+namespace mozilla::dom {
+
+bool ServiceWorkerLifetimeExtension::LifetimeExtendsIntoTheFuture(
+    uint32_t aRequiredSecs) const {
+  // Figure out the new deadline, a null TimeStamp means no change.
+  return this->match(
+      [](const NoLifetimeExtension& nle) { return false; },
+      [aRequiredSecs](const PropagatedLifetimeExtension& ple) {
+        TimeStamp minFuture =
+            TimeStamp::NowLoRes() + TimeDuration::FromSeconds(aRequiredSecs);
+
+        // Ignore null deadlines or deadlines that don't extend sufficiently
+        // into the future.
+        return !(ple.mDeadline.IsNull() || ple.mDeadline < minFuture);
+      },
+      [](const FullLifetimeExtension& fle) { return true; });
+}
+
+}  // namespace mozilla::dom